src/validations/validations.ts

```diff
@@ -44,96 +73,112 @@ export function validateMetatile(metatile: number): void {
   validateBoundingBox(boundingBox);
 
-  validateGeoPointByGrid(boundingBox.min, referenceTileGrid);
-  validateGeoPointByGrid(boundingBox.max, referenceTileGrid);
+  try {
+    validateGeoPointByTileMatrix(boundingBox.min, tileMatrix);
+    validateGeoPointByTileMatrix(boundingBox.max, tileMatrix);
+  } catch (err) {
+    throw new RangeError(`bounding box is not within the tile matrix`);
+  }
 }
 
 /**
- * Validates that the input `geometry` is a valid geometry
- * @param geometry
- * @param referenceTileGrid
+ * Validates that the input `geometry` is a valid with respect to `tileMatrixSet`
+ * @param geometry geometry
+ * @param tileMatrix tile matrix
  */
-export function validateGeometryByGrid<G extends GeoJSONGeometry>(geometry: Geometry<G>, referenceTileGrid: TileGrid): void {
+export function validateGeometryByTileMatrix<G extends GeoJSONGeometry>(geometry: Geometry<G>, tileMatrix: TileMatrix): void {
   const boundingBox = geometryToBoundingBox(geometry);
-  validateBoundingBoxByGrid(boundingBox, referenceTileGrid);
+  validateBoundingBoxByTileMatrix(boundingBox, tileMatrix);
 }
 
 /**
- * Validates that the input `tile` is valid
- * @param tile the tile to validate
- * @param referenceTileGrid the tile grid to validate the `tile` against
+ * Validates that the input `tile` is valid with respect to `tileMatrixSet`
+ * @param tile tile to validate
+ * @param tileMatrixSet tile matrix set to validate the `tile` against
  */
-export function validateTileByGrid(tile: Tile, referenceTileGrid: TileGrid): void {
-  validateZoomByGrid(tile.z, referenceTileGrid);
-  if (tile.metatile !== undefined) {
-    validateMetatile(tile.metatile);
+export function validateTileByTileMatrix<T extends TileMatrixSet>(tile: Tile<T>, tileMatrix: ArrayElement<T['tileMatrices']>): void {
+  const { col, row, tileMatrixId, metatile } = tile;
+  if (metatile !== undefined) {
+    validateMetatile(metatile);
+  }
+
+  if (tileMatrixId !== tileMatrix.identifier.code) {
+    throw new Error('tile identifier is not equal to the tile matrix identifier');
+  }
+
+  if (col < 0 || col >= tileMatrix.matrixWidth / (metatile ?? 1)) {
+    throw new RangeError('tile matrix col index out of range of the tile matrix');
+  }
+
+  if (row < 0 || row >= tileMatrix.matrixHeight / (metatile ?? 1)) {
+    throw new RangeError('tile matrix row index out of range of the tile matrix');
+  }
+}
+
+/**
+ * Validates that the input `tile` is valid with respect to `tileMatrixSet`
+ * @param tile tile to validate
+ * @param tileMatrixSet tile matrix set to validate the `tile` against
+ */
+export function validateTileByTileMatrixSet<T extends TileMatrixSet>(tile: Tile<T>, tileMatrixSet: T): void {
+  const { col, row, tileMatrixId, metatile } = tile;
+  if (metatile !== undefined) {
+    validateMetatile(metatile);
+  }
+
+  const tileMatrix = getTileMatrix(tileMatrixId, tileMatrixSet);
+  if (!tileMatrix) {
+    throw new Error('tile could not be found inside tile matrix set');
   }
```

Comment:
Why not call `validateTileByTileMatrix` after this?

Reply:
it was because of the more precise error messages
fixing